specifyweb/frontend/js_src/lib/components/WorkBench/CellMeta.ts

```diff
@@ -46,6 +46,386 @@ const defaultMetaValues = Object.freeze([
+      initialVisualRow ?? this.workbench.hot.toVisualRow(physicalRow);
+    const visualCol =
+      initialVisualCol ?? this.workbench.hot.toVisualColumn(physicalCol);
+    const cell = initialCell ?? this.workbench.hot.getCell(visualRow, visualCol);
+    this.runMetaUpdateEffects(
+      cell ?? undefined,
+      key,
+      value,
+      visualRow,
+      visualCol
+    );
+
+    return true;
+  }
+
+  /*
+   * Returns this.cellMeta, but instead of indexing by physical row/col,
+   * indexes by visual row/col
+   *
+   * This is used for navigation among cells.
+   *
+   * Also, if navigation direction is set to ColByCol, the resulting array
+   * is transposed.
+   *
+   * this.flushIndexedCellData is set to true whenever visual indexes change
+   *
+   */
+  public getCellMetaObject(): RA<RA<WbMetaArray>> {
+    if (this.flushIndexedCellData || this.indexedCellMeta === undefined) {
+      if (this.workbench.hot === undefined) return [];
+      const resolveIndex = (
+        visualRow: number,
+        visualCol: number,
+        first: boolean
+      ) =>
+        (this.workbench.searchPreferences.navigation.direction ===
+          'rowFirst') ===
+        first
+          ? visualRow
+          : visualCol;
+
+      const indexedCellMeta: WritableArray<WritableArray<WbMetaArray>> = [];
+      Object.entries(this.cellMeta).forEach(([physicalRow, metaRow]) =>
+        Object.entries(metaRow).forEach(([physicalCol, cellMeta]) => {
+          const visualRow = this.workbench.hot!.toVisualRow(
+            (physicalRow as unknown as number) | 0
+          );
+          const visualCol = this.workbench.hot!.toVisualColumn(
+            (physicalCol as unknown as number) | 0
+          );
+          indexedCellMeta[resolveIndex(visualRow, visualCol, true)] ??= [];
+          indexedCellMeta[resolveIndex(visualRow, visualCol, true)][
+            resolveIndex(visualRow, visualCol, false)
+          ] = cellMeta;
+        })
+      );
+      this.indexedCellMeta = indexedCellMeta;
+    }
+    this.flushIndexedCellData = false;
+    return this.indexedCellMeta;
+  }
+
+  // MetaData
+  updateCellInfoStats() {
+    const cellMeta = this.cellMeta.flat();
+
+    this.cellCounts = {
+      newCells: cellMeta.reduce(
+        (count, info) =>
+          count + (this.getCellMetaFromArray(info, 'isNew') ? 1 : 0),
+        0
+      ),
+      invalidCells: cellMeta.reduce(
+        (count, info) =>
+          count +
+          (this.getCellMetaFromArray(info, 'issues').length > 0 ? 1 : 0),
+        0
+      ),
+      searchResults: cellMeta.reduce(
+        (count, info) =>
+          count + (this.getCellMetaFromArray(info, 'isSearchResult') ? 1 : 0),
+        0
+      ),
+      modifiedCells: cellMeta.reduce(
+        (count, info) =>
+          count + (this.getCellMetaFromArray(info, 'isModified') ? 1 : 0),
+        0
+      ),
+    };
+    
+    // below line breaks error cell highlighting
+    // this.workbench.setCellCounts(this.cellCounts);
```

Comment:
Was using `this.workbench.setCellCounts(this.cellCounts)` to set cellCounts for the `Navigation` component on the bottom. However, this breaks error cell highlighting completely. Without having a state for cellCounts, there isn't a good way to render the cell count in the navigation buttons. Need to find a way to make my components work with the vanilla classes we have.